cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/LocalApplicationTemplateManager.java

```diff
@@ -0,0 +1,119 @@
+  private final CConfiguration configuration;
+  private final Store store;
+  private final StreamConsumerFactory streamConsumerFactory;
+  private final QueueAdmin queueAdmin;
+  private final DiscoveryServiceClient discoveryServiceClient;
+  private final StreamAdmin streamAdmin;
+  private final ExploreFacade exploreFacade;
+  private final boolean exploreEnabled;
+  private final AdapterService adapterService;
+  private final ProgramTerminator programTerminator;
+  private final DatasetFramework datasetFramework;
+  private final DatasetFramework inMemoryDatasetFramework;
+  private final PreferencesStore preferencesStore;
+
+  @Inject
+  public LocalApplicationTemplateManager(CConfiguration configuration, PipelineFactory pipelineFactory,
+                                         NamespacedLocationFactory namespacedLocationFactory,
+                                         Store store, StreamConsumerFactory streamConsumerFactory,
+                                         QueueAdmin queueAdmin, DiscoveryServiceClient discoveryServiceClient,
+                                         DatasetFramework datasetFramework,
+                                         @Named("datasetMDS") DatasetFramework inMemoryDatasetFramework,
+                                         StreamAdmin streamAdmin, ExploreFacade exploreFacade,
+                                         AdapterService adapterService,
+                                         PreferencesStore preferencesStore,
+                                         @Assisted ProgramTerminator programTerminator) {
+    this.configuration = configuration;
+    this.namespacedLocationFactory = namespacedLocationFactory;
+    this.pipelineFactory = pipelineFactory;
+    this.discoveryServiceClient = discoveryServiceClient;
+    this.store = store;
+    this.streamConsumerFactory = streamConsumerFactory;
+    this.queueAdmin = queueAdmin;
+    this.programTerminator = programTerminator;
+    this.datasetFramework = datasetFramework;
+    this.inMemoryDatasetFramework = inMemoryDatasetFramework;
+    this.streamAdmin = streamAdmin;
+    this.exploreFacade = exploreFacade;
+    this.exploreEnabled = configuration.getBoolean(Constants.Explore.EXPLORE_ENABLED);
+    this.adapterService = adapterService;
+    this.preferencesStore = preferencesStore;
+  }
+
+  @Override
+  public ListenableFuture<ApplicationWithPrograms> deploy(Id.Namespace id, @Nullable String templateId,
+                                                          DeploymentInfo input) throws Exception {
+    Pipeline<ApplicationWithPrograms> pipeline = pipelineFactory.getPipeline();
+    pipeline.addLast(new LocalArchiveLoaderStage(store, configuration, id, templateId));
+    pipeline.addLast(new ApplicationTemplateVerificationStage(store, datasetFramework, adapterService));
+    pipeline.addLast(new DeployDatasetModulesStage(configuration, datasetFramework, inMemoryDatasetFramework));
```

Comment:
Was thinking there could be a case for having a template wide dataset or something like that. Kind of like a static variable in a class.  But maybe it would be best to restrict it unless we have a hard use case, so i will take it out.